test_quality_of_service/test/test_deadline.cpp

```diff
@@ -0,0 +1,121 @@
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <tuple>
+
+#include "gtest/gtest.h"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executors/single_threaded_executor.hpp"
+
+#include "std_msgs/msg/string.hpp"
+
+#include "test_quality_of_service/qos_utilities.hpp"
+
+
+using namespace std::chrono_literals;
+
+/// Test Deadline with a single publishing node and single subscriber node
+TEST_F(TestSetup, test_deadline) {
+  int expected_number_of_pub_events = 5;
+  const std::chrono::milliseconds deadline_duration = 1s;
+  std::tuple<size_t, size_t> deadline_duration_tuple = chrono_milliseconds_to_size_t(
+    deadline_duration);
+  const std::chrono::milliseconds max_test_length = 10s;
+  const std::chrono::milliseconds publish_rate = deadline_duration / 5;
+
+  // used for lambda capture
+  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> exec = executor;
+  int total_number_of_publisher_deadline_events = 0;
+  int total_number_of_subscriber_deadline_events = 0;
+  int last_pub_count = 0;
+  int last_sub_count = 0;
+
+  rmw_qos_profile_t qos_profile = rmw_qos_profile_default;
+  qos_profile.deadline.sec = std::get<0>(deadline_duration_tuple);
+  qos_profile.deadline.nsec = std::get<1>(deadline_duration_tuple);
+
+  // setup subscription options and callback
+  rclcpp::SubscriptionOptions<> subscriber_options;
+  subscriber_options.qos_profile = qos_profile;
+  subscriber_options.event_callbacks.deadline_callback =
+    [&last_pub_count,
+    &total_number_of_subscriber_deadline_events](rclcpp::QOSDeadlineRequestedInfo & event) -> void
+    {
+      RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"), "QOSDeadlineRequestedInfo callback");
+      total_number_of_subscriber_deadline_events++;
+      // assert the correct value on a change
+      ASSERT_EQ(1, event.total_count_change);
+      last_pub_count = event.total_count;
+    };
+
+  // setup publishing options and callback
+  rclcpp::PublisherOptions<> publisher_options;
+  publisher_options.qos_profile = qos_profile;
+  publisher_options.event_callbacks.deadline_callback =
+    [&last_sub_count,
+    &total_number_of_publisher_deadline_events](rclcpp::QOSDeadlineOfferedInfo & event) -> void
+    {
+      RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"), "QOSDeadlineOfferedInfo callback");
+      total_number_of_publisher_deadline_events++;
+      // assert the correct value on a change
+      ASSERT_EQ(1, event.total_count_change);
+      last_sub_count = event.total_count;
+    };
+
+  std::string topic("test_deadline");
+
+  auto publisher = std::make_shared<Publisher>("publisher", topic, publisher_options, publish_rate);
+  auto subscriber = std::make_shared<Subscriber>("subscriber", topic, subscriber_options);
+
+  // toggle publishing on and off to force deadline events
+  rclcpp::TimerBase::SharedPtr kill_publisher_timer = nullptr;
+  kill_publisher_timer = subscriber->create_wall_timer(
+    max_test_length / expected_number_of_pub_events,
+    [&publisher]() -> void {
```

Comment:
I would advise capturing by copy here. This code is susceptible to the following race:

```
1. publisher.reset()
2. Callback is called concurrently
3. publisher contains a random value => CRASH
```

If you want to keep this as a ref, you need a lock unless you can ensure that the publisher value will never be touched in the main thread before the timer is cancelled which is hard to achieve (we can get it right for this PR, but next change to this file will probably fail to remember this nuance).

Reply:
Found some ref for that:
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rf-value-capture

Reply:
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rconc-data-by-value

Reply:
A copy won't work as the publisher has to be explicitly reset in order to generate the liveliness event. Will have to use a lock.